HTML-642 and HTML-643 added MetadataMappingService

AdamGrzybkowski opened the pull request.

mogoodrich commented:
I just took at quick look at this, but it looks good.   Looks like it is properly designed to only be metamapping module "aware" but we probably should double-check that everything still runs with the metadatamapping module not installed, for what it's worth.

@rkorytkowski , does this look good to you?

dkayiwa commented:
@rkorytkowski did you see @mogoodrich's question above?

rkorytkowski commented on api-tests/src/test/resources/TestingApplicationContext.xml:
```diff
@@ -20,6 +20,12 @@
+        <property name="useSystemClassLoader">
+            <value>true</value>
+        </property>
+    </bean>
```
Why is it here?

rkorytkowski commented on api/src/main/resources/moduleApplicationContext.xml:
```diff
@@ -6,6 +6,18 @@
+                </bean>
+            </list>
+        </property>
+    </bean>
```
Doesn't it break the conditional loading of MetadataMappingServiceWrapper? Have you tested it running the module without the metadata mapping module? Also MetadataMappingServiceWrapper doesn't extend OpenmrsService... and classes passed to moduleService are expected to be OpenmrsServices. I would remove it here and get the bean using Context.getRegisteredComponent() instead of Context.getService()

rkorytkowski commented on api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormEntryUtil.java:
```diff
@@ -1139,6 +1166,11 @@ public static ProgramWorkflowState getState(String identifier) {
+
+			int index = identifier.indexOf(":");
+			if(index != -1){
+				return getMetadata(identifier, ProgramWorkflowState.class, index);
```
This feels strange when reading... I would change this bit of code to:
ProgramWorkflowState state = getMetadataByMapping(ProgramWorkflowState.class, identifier);
if (state != null) {
return state;
}

rkorytkowski commented on api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormExporter.java:
```diff
@@ -136,7 +138,20 @@ private void calculateDependencies() {
+										int index = id.indexOf(":");
+										String source = id.substring(0, index);
+										String code = id.substring(index+1, id.length());
+										OpenmrsObject object = metadataMappingServiceWrapper.getMetadataItem((Class<? extends OpenmrsMetadata>) attributeDescriptor.getClazz(), source, code);
```
It feels like the logic of splitting the identifier belongs to the metadataMappingServiceWrapper... Basically only the metadataMappingServiceWrapper should know what is the format of the identifier and if it is the correct format then do the search. We don't want to see ":" all over the code :)

AdamGrzybkowski commented:
@rkorytkowski  I've updated the code but I have to following error now. 
When running tests in Intellij everything passes, but when I try to do mvn clean install there are tests errors. It looks like the metadataMappingServiceWrapper is not injected, but can't figure out why.
I have tried to edit maven surefire plugin config and changed forkMode to 'always' and my tests seems to pass but the EnrollInProgramTagTest fails.

AdamGrzybkowski commented:
I couldn't use the OpenmrsProfile to load metadata mapping conditionally, because there was no way to test it.

dkayiwa commented:
@AdamGrzybkowski so what did you end up using? :)

dkayiwa commented:
@AdamGrzybkowski did you notice the merge conflicts on this pull request?

dkayiwa commented:
@AdamGrzybkowski did you abandon this? :)

dkayiwa commented:
@AdamGrzybkowski feel free to reopen this whenever you are ready to resume. :)

AdamGrzybkowski commented:
@dkayiwa I didn't get any notification from this PR. This PR was ready when I added my last comment.If you still want this code, please reopen this PR :) I will resolve conflicts

PS. Did you add me to openMrs organization on GitHub ?

dkayiwa commented:
@AdamGrzybkowski i sent you an invitation. Status = Pending (waiting for you to accept) :)

AdamGrzybkowski commented:
@dkayiwa Conflicts resolved :) 

Sorry for the late response, but I don't get email notifications and I forgot about this PR :)

mogoodrich commented:
@rkorytkowski  does this look good now?  If so, feel free to merge.  Thanks!

dkayiwa commented:
@mogoodrich Rafal is on a 2 weeks vacation. :)

mogoodrich commented:
Ah, okay, I will go ahead and merge then!

mogoodrich commented:
Though maybe I should have waited until Bamboo was working again... the PIH build pipeline is testing it at least.
